code/modules/hydroponics/biogenerator.dm

```diff
@@ -1,337 +1,354 @@
-	if(materials.len != 1 || materials[1] != MAT_BIOMASS)
-		return 0
-	if (materials[MAT_BIOMASS]*multiplier/efficiency > points)
-		menustat = "nopoints"
-		return 0
-	else
-		if(remove_points)
-			points -= materials[MAT_BIOMASS]*multiplier/efficiency
-		update_icon()
-		updateUsrDialog()
-		return 1
+			biomass += (plant.reagents.get_reagent_amount("nutriment") + plant.reagents.get_reagent_amount("plantmatter")) * 10 * productivity
+		qdel(plant)
+
+	playsound(loc, 'sound/machines/blender.ogg', 50, 1)
+	use_power(S * 30)
+	sleep(S + 15 / productivity)
+	stored_plants = list()
+	processing = FALSE
+	SStgui.update_uis(src)
+	update_icon()
 
-/obj/machinery/biogenerator/proc/check_container_volume(list/reagents, multiplier = 1)
+/**
+ * Ejects the biogenerator's stored plants
+ */
+/obj/machinery/biogenerator/proc/eject_plants()
+	for(var/obj/plant as anything in stored_plants)
+		plant.forceMove(get_turf(src))
+	stored_plants = list()
+	SStgui.update_uis(src)
+
+/**
+ * Check if the biogenerator has enough biomass to create the passed in design `D`, times the `mutiplier`.
+ *
+ * Arguments:
+ * * datum/design/D - the design we want to create
+ * * multiplier - how many of this design should be built
+ */
+/obj/machinery/biogenerator/proc/check_cost(datum/design/D, multiplier = 1)
+	if(!D.materials[MAT_BIOMASS] || ((D.materials[MAT_BIOMASS] / efficiency) * multiplier) > biomass)
+		return FALSE
+	return TRUE
+
+/**
+ * Check if the biogenerator's `container` can hold the reagents we're trying to give it.
+ *
+ * Arguments:
+ * * datum/design/D - the design we want to create reagents from
+ * * multiplier - how many of this design should be built
+ */
+/obj/machinery/biogenerator/proc/check_container_volume(datum/design/D, multiplier = 1)
 	var/sum_reagents = 0
-	for(var/R in reagents)
-		sum_reagents += reagents[R]
+	for(var/R in D.make_reagents)
+		sum_reagents +=  D.make_reagents[R]
 	sum_reagents *= multiplier
 
-	if(beaker.reagents.total_volume + sum_reagents > beaker.reagents.maximum_volume)
-		menustat = "nobeakerspace"
-		return 0
-
-	return 1
-
+	if(container.reagents.total_volume + sum_reagents > container.reagents.maximum_volume)
+		return FALSE
+	return TRUE
+
+/**
+ * Create biogenerator products and subtracts the appropriate biomass cost.
+ *
+ * Arguments:
+ * * datum/design/D - the design we want to create, or create reagents from
+ * * amount - how many of this design should be built
+ */
 /obj/machinery/biogenerator/proc/create_product(datum/design/D, amount)
-	if(!beaker || !loc)
-		return 0
-
+	// Creating stack-based items like cloth or cardboard.
 	if(ispath(D.build_path, /obj/item/stack))
-		if(!check_container_volume(D.make_reagents, amount))
-			return 0
-		if(!check_cost(D.materials, amount))
-			return 0
-
-		var/obj/item/stack/product = new D.build_path(loc)
+		if(!check_cost(D, amount))
+			return
+		var/obj/item/stack/product = new D.build_path(get_turf(src))
 		product.amount = amount
```

Comment:
You can pass in the amount right to `stack/New()`, but idk if that's more readable..